app/questionnaire/navigator.py

```diff
@@ -67,13 +64,13 @@ def build_path(self, blocks, this_location):
 
             # If routing rules exist then a rule must match (i.e. default goto)
             if 'routing_rules' in block and len(block['routing_rules']) > 0:
-                original_this_location = this_location.copy()
+                original_this_location = copy.copy(this_location)
                 for rule in filter(is_goto_rule, block['routing_rules']):
-                    should_goto = evaluate_goto(rule['goto'], self.metadata, self.answer_store, this_location['group_instance'])
+                    should_goto = evaluate_goto(rule['goto'], self.metadata, self.answer_store, this_location.group_instance)
 
                     if should_goto:
-                        next_location = this_location.copy()
-                        next_location['block_id'] = rule['goto']['id']
+                        next_location = copy.copy(this_location)
```

Comment:
We can remove this copy now and just modify `this_location` directly as you can use `original_this_location` on line 79 to do the remove, this eliminates the need for `next_location`